src/Microsoft.Diagnostics.Monitoring.WebApi/ActionContextExtensions.cs

```diff
@@ -91,4 +83,43 @@ private static bool LogInformation(ILogger logger, Exception ex)
             return true;
         }
     }
+
+    internal static class ActionContextExtensions
+    {
+        public static Task ProblemAsync(this ActionContext context, BadRequestObjectResult result)
+        {
+            if (context.HttpContext.Features.Get<IHttpResponseFeature>().HasStarted)
+            {
+                // If already started writing response, do not rewrite
+                // as this will throw an InvalidOperationException.
+                return Task.CompletedTask;
+            }
+            else
+            {
+                // Need to manually add this here because this ObjectResult is not processed by the filter pipeline,
+                // which would look up the applicable content types and apply them before executing the result.
+                result.ContentTypes.Add(ContentTypes.ApplicationProblemJson);
+                return result.ExecuteResultAsync(context);
+            }
+        }
+
+        public static async Task InvokeAsync(this ActionContext context, Func<CancellationToken, Task> action, ILogger logger)
+        {
+            //We want to handle two scenarios:
+            //1) These functions are part of an ExecuteResultAsync implementation, in which case
+            //they don't need to return anything. (The delegate below handles this case).
+            //2) They are part of deferred processing and return a result.
+            Func<CancellationToken, Task<ExecutionResult<object>>> innerAction = async (CancellationToken token) =>
+            {
+                await action(token);
+                return new ExecutionResult<object>();
```

Comment:
nit: Maybe create a static instance and return that so long as ExecutionResult is immutable.